pkg/kubefed2/federate/federate.go

```diff
@@ -0,0 +1,281 @@
+	targetClient, err := ctlutil.NewResourceClient(hostConfig, &targetAPIResource)
+	if err != nil {
+		return nil, errors.Wrap(err, "Error creating target resource client")
+	}
+
+	resource, err := targetClient.Resources(qualifiedResourceName.Namespace).Get(qualifiedResourceName.Name, metav1.GetOptions{})
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil, errors.New(fmt.Sprintf("Target resource %q not found", qualifiedResourceName))
+		}
+		return nil, errors.Wrap(err, "Error retreiving the target resource")
+	}
+
+	return resource, nil
+}
+
+func createFedResource(hostConfig *rest.Config, typeConfig *fedv1a1.FederatedTypeConfig, template *unstructured.Unstructured) error {
+	fedAPIResource := typeConfig.GetFederatedType()
+	fedClient, err := ctlutil.NewResourceClient(hostConfig, &fedAPIResource)
+	if err != nil {
+		return errors.Wrap(err, "Error creating federation resource client")
+	}
+
+	RemoveControllerSetFields(template)
+	resourceNameSpace := template.GetNamespace()
+	resourceName := template.GetName()
+	fedResource := &unstructured.Unstructured{}
+	SetBasicMetaFields(fedResource, fedAPIResource, resourceName, resourceNameSpace, "")
+
+	err = unstructured.SetNestedField(fedResource.Object, template.Object, ctlutil.SpecField, ctlutil.TemplateField)
+	if err != nil {
+		return errors.Wrap(err, "Error setting template into Federated resource")
+	}
+
+	readyClusters, err := getReadyClusters(hostConfig, typeConfig.Namespace)
+	if err != nil {
+		return err
+	}
+	err = ctlutil.SetClusterNames(fedResource, readyClusters)
+	if err != nil {
+		return errors.Wrap(err, "Error setting cluster names into Federated resource")
+	}
+
+	_, err = fedClient.Resources(resourceNameSpace).Create(fedResource, metav1.CreateOptions{})
+	if err != nil {
+		return errors.Wrap(err, "Error creating Federate resource")
+	}
+
+	glog.Infof("Federated resource: %s/%s of type referenced by typeconfig: %s/%s created.", resourceNameSpace, resourceName, typeConfig.Namespace, typeConfig.Name)
+	return nil
+}
+
+// TODO(irfanurrehman): figure out a better way of doing this
+func RemoveControllerSetFields(resource *unstructured.Unstructured) {
+	systemMetadataFields := []string{"selfLink", "uid", "resourceVersion", "generation", "creationTimestamp", "deletionTimestamp", "deletionGracePeriodSeconds"}
+	for _, field := range systemMetadataFields {
+		unstructured.RemoveNestedField(resource.Object, ctlutil.MetadataField, field)
+		// For resources with pod template subresource (jobs, deployments, replicasets)
+		unstructured.RemoveNestedField(resource.Object, ctlutil.SpecField, ctlutil.TemplateField, ctlutil.MetadataField, field)
+	}
+}
+
+func SetBasicMetaFields(resource *unstructured.Unstructured, apiResource metav1.APIResource, name, namespace, generateName string) {
+	resource.SetKind(apiResource.Kind)
+	gv := schema.GroupVersion{Group: apiResource.Group, Version: apiResource.Version}
+	resource.SetAPIVersion(gv.String())
+	if name != "" {
```

Comment:
Why is it necessary to perform this check vs just blindly setting the field?

Reply:
Now setting without check!